STL_Extension/include/CGAL/Uncertain.h

```diff
@@ -520,6 +494,69 @@ Uncertain<T> make_uncertain(Uncertain<T> t)
   return t;
 }
 
+// operator&& and operator|| are not provided because, unless their bool counterpart,
+// they lack the "short-circuiting" property.
+// We provide macros CGAL_AND and CGAL_OR, which attempt to emulate their behavior.
+// Key things : do not evaluate expressions twice, and evaluate the right hand side
+// expression only when needed.
+
+inline
+Uncertain<bool> CGAL_AND(const Uncertain<bool>& a, const Uncertain<bool>& b)
+{
+  return certainly_not(a) ? make_uncertain(false)
+                          : a & b;
```

Comment:
That looks equivalent to just `a&b` and doesn't provide the short-circuiting behavior that was the whole reason for using statement expressions. You could use this to replace the `&&` fallback for CGAL_CFG_NO_STATEMENT_EXPRESSIONS, but not for gcc/clang.
As indicated in the old comment, you could try using lambdas, like (completely untested)
```c++
#define CGAL_AND(A,B) plouf((A),[&](){return (B);})
template<class BF>
auto plouf(Uncertain<bool> a, BF&&bf){
  if(certainly_not(a))return make_uncertain(false);
  return a & make_uncertain(bf());
}
```

Reply:
I don't understand why the code I wrote is not equivalent to the old one. I'm probably missing something but to me the goal was to evaluate `a&b` only if a is not certainly false.

Reply:
Currently, `CGAL_AND(false, f())` does not call `f` at all with gcc/clang. And `f` may be some rather costly predicate... Evaluating `a&b` is cheap if `a` and `b` are already known, the point is to avoid computing b, just like `false&&f()` does not call `f`. In the code you wrote, you always compute both `a` and `b` to pass them to the function. In the macro code, the computation of `b` is in a branch that is not always taken.

Reply:
oh I see, b might be the result of a function directly passed to the macro.